### PR TITLE
[FLINK-19457][core] Add a number sequence generating source for the New Source API.

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This adds the `NumberSequenceSource` that generates a data stream of numbers. This is a useful source for testing, debugging, or to simply have a stream of placeholder events.

A number-sequence source currently exists in the form of `env.generateSequence(from to)` based on the legacy Source API.
The source added here should in a later commit replace that implementation (the `StatefuSequenceSource`).

## Verifying this change

This test adds a simple unit test for the checkpoint/restore functionality. 
The source is used in various integration tests already, we hence do not add an additional end-to-end test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no** (replaces existing)
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java:48]:
```suggestion
 * A data source that produces a sequence of numbers (longs).
```

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java:74]:
```suggestion
	/**
	 * Creates a new {@code NumberSequenceSource} that produces parallel sequences covering the range
	 * {@code from} to {@code to} (both boundaries are inclusive).
	 */
```

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceEnumerator.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceEnumerator.java:34]: I would use `{@link SplitEnumerator}` because it will create links and allows following them in the IDE.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java:41]:
```suggestion
 * A {@link SourceReader} that returns the values of an iterator, supplied via an
```

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java:62]: nit: broken formatting

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java:67]: nit: broken formatting

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java:97]: nit: the newline style for the `if-else` branches is inconsistent here

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceSplit.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceSplit.java:26]:
```suggestion
 * A {@link SourceSplit} that represents a sequence of elements captured in an iterator.
```

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java:26]:
```suggestion
 * A {@link SourceEvent} indicating that there is no split available for the reader (any more).
```

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java:26]:
```suggestion
 * A {@link SourceEvent} representing the request for a split, typically sent from the
```

**Review by aljoscha — COMMENTED** [review]
I looked at this mostly to educate myself but to me it looks good. Maybe @becketqin will have more to say.

I did have some nitpicks about formatting and Javadocs.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java
Comment (becketqin) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java:191]: Can we add a comment to explain the number of 16 here?

**Comment by StephanEwen** [comment]
Thanks, @aljoscha, these suggestions sound good to me.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java
Comment (becketqin) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java:227]: The number of 20 is a little confusing. Can we add a comment here?

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java
Comment (becketqin) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java:29]: Just want to confirm, we don't need this class any more after [FLINK-19265](https://issues.apache.org/jira/browse/FLINK-19265), right?

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java
Comment (becketqin) [inline flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java:29]: There is also a `RequestSplitEvent` in `SourceReaderBase`. I am a little unsure if that should be reused. On one hand, it is just an event, the cost of duplicating is little. On the other hand, they are for the exact same purpose, so ideally we want just one class. That means the implementation should depend on `flink-connector-base`, which is doable if we move this source implementation to `flink-connector-testing` as mentioned in the top level comment. 

What do you think?

**Review by becketqin — COMMENTED** [review]
@StephanEwen Thanks for the patch. It is definitely useful for testing and playing around.

Regarding the packaging, do you think it would be better to have a `flink-connectors/flink-connector-testing` or `flink-connectors/flink-connector-utils` module? Technically speaking this is a specific implementation instead of the Source API perse. Having that in `flink-core` seems a little confusing.

File: flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java
Comment (becketqin) [inline flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java:39]: Should we also have an IT case for the source as well? This would help reveal issues earlier if this source is intended to be used by other test cases.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java
Comment (StephanEwen) [reply flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java:29]: I would not move this to a testing project, because this source is used in non-testing as well.
The predecessor of this is even directly integrated in the DataStream API at the moment: `env.generateSequence()`

File: flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java
Comment (StephanEwen) [reply flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java:39]: I was contemplating this and decided against that, as mentioned in the PR description.

The Integration Test would mostly test the runtime source operator and coordinator, not specific to this source here.
Having at least one case of the source being used end to end makes total sense, and this is the case in other tests (see for example #13513)

Comment (StephanEwen) [reply flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java:39]: Adding an ITCase is certainly possible, it would not be much work, it would just be added testing time for something that is already covered.

**Comment by StephanEwen** [comment]
@becketqin We used to have various "no dependency" built-in sources together with the source APIs so far.
Check out `InputFormat` and `SourceFunction` - there was always a lot of concrete implementations directly in the project, if they were of the "utility kind" with no dependencies. Similar as `java.util` has the collection framework interfaces, but also the simple implementations (that are no concurrent, distributed, etc.).

I feel having a separate module and dependency for those is a bit heavy-weight.
It would also make it impossible to "deep integrate" these sources with the API, like with the convenience methods we currently have (even though that is optional to do, in my mind).

**Comment by becketqin** [comment]
@StephanEwen Personally speaking, I think usually people either put both API and implementation in the same package or completely separate them. The weird part here is that we have an explicit `api` package containing concrete implementation. Should we at least take that out of the `org.apache.flink.api` package, maybe to `org.apache.flink.util.source` in `flink-core`?

Regarding "deep integrate", I feel that this is double-edged and needs to be considered case by case. I was actually wondering why a `StreamExecutionEnvironment` also creates sources while it is supposed to take sources to generate `DataStream`. I feel `env.addSource(ConnectorUtils.fromCollection())` is equally convenient, where `ConnectorUtils` looks like something below

```
ConnectorUtils {

    NumSequenceSource generateSequence(long from, long to);

    <T> CollectionSource<T> fromCollection(Collection<T> collections);

    BlackHoleSink blackHoleSink();

    PrintingSink printSink();

    ...
}
```

 It follows the general design principle of one class only does one thing, without overlapping in responsibilities.  `StreamExecutionEnv#addSource()` is the only entrance of registering sources to create new `DataStreams`. `ConnectorUtils` is the only entrance of getting builtin connectors. Conceptually, I feel this is easier for users to follow than having `deep integrate` in separate classes.

File: flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java
Comment (becketqin) [reply flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java:39]: It seems that we should not reduce test time by skipping reasonable tests? The concern I has was that if there is another test using the `NumberSequenceSource` failed due to a bug in this source. The author of that test may have to look into the code of this source, which seems something we would like to avoid by having tests here.

**Comment by StephanEwen** [comment]
For both topics, this PR here stays close to what the previous state of sources was, in an effort to port the previous state to the new API.

We can open the discussions about having different package structures in the future, but do we need to block this PR on that?
The package structure in flink-core has never separated API and implementation, and the `api` name was not chosen with very much consideration. IIRC mainly to separate SDK from runtime parts. (Arguable `.sdk.` would have been the correct name, but were were all younger and less experienced when we created these parts.)

From the user's perspective I think it is better to stay consistent within one style then have different packages follow different styles. That adds more confusion and makes it harder to navigate.

Similar for the "deep integration", I feel this should not block this PR (foremost because this PR does not touch anything there). 
That is a different discussion that we should either apply everywhere or not do it. Doing this now here differently than in all other places will add confusion for users. Another side of the problem is that many of these methods `env.generateSequence(...)` are public API that we should not break.

**Comment by becketqin** [comment]
@StephanEwen Thanks for the explanation. I checked the classes and structure in `flink-core` again, I agree that this PR is in general following the existing convention. And I am fine with merge the PR as is. 

That said, I had a hard time to fully understand the structure of `flink-core`. It would be great if you can help me understand this a bit more. I am wondering if the users would have the same questions like below.

- I was initially expecting the function interfaces in `org.apache.api.java.functions` path, but there are only `KeySelector` and `NullByteKeySelector`. The actual function interfaces are in `org.apache.flink.common.functions` package. What is the difference between those two packages?
- Also, I was not quite sure about the differences between `org.apache.flink.api.common.io` and `org.apache.flink.core.io`. 
- In the existing package paths in `flink-core`, the implementation and interface classes are put in the same package. This PR introduces a new package `lib`. Should the other implementations also be put in a `lib` path?
- In the existing package paths, `util` packages usually contains internal helper classes for the (abstract) implementations of the interfaces. This PR seems putting the abstract implementations of the `SourceReader` / `Enumerator` interfaces in the `util` package, and expecting other Iterator based sources to leverage that? Or are those classes also internal?
- The follow up question to the above one, it seems that the `IteratorSourceReader`/ `IteratorSourceEnumerator` classes are just another base implementation of the `Source` interface, why the base implementation of the source in `flink-connector-base` was not put into `org.apache.flink.api.connector.source.lib.util`? 

Sorry for sticking to discussion, my main concern is that If we ever want to revisit the API and package structure, right now might be the best timing, because we are essentially introducing new `Source` APIs and will probably deprecate old `SourceFunction` and `InputFormat` APIs in some coming release. This would be a natural migration towards new APIs if we want to do that. Keeping the API / package structure as is and changing them in the future could be even more disruptive.

Thanks again.

**Comment by StephanEwen** [comment]
No worries, let's figure this out together.

### About the flink-core package structure.

Agree, `flink-core`'s package structure is wild.

Most of it is because of the pre-streaming origin, and there was not too much thought there. `api.common` was shared between Batch Java and Batch Scala APIs. `api.java` was originally in a different project (Batch Java API) and moved to core when streaming came into the mix to be shared with the Streaming Java API.

`core.io` was IIRC IO-related stuff that was not part of the API, but indirectly relevant in core API classes, like the utils around the managed memory abstraction. `api.common.io` was for API-related I/O, like sources, sinks.

A lot of stuff was also added over time in the wrong places, admittedly, because few people in the early days paid attention to package structure. The packages stuck because we didn't want to break user applications. For Flink 2.0, we should try and simplify this.

tl;dr - it is the way it is for historic reasons and because the "not breaking application" consideration ended up trumping the "clean up packages" consideration. Seems the structure was not getting too much in the way of users after all.

### "flink-connector-base" versus "flink-core"

What I would like to keep is that this source implementation stays in `flink-core` and does not move to another module. We have a lot of dependencies already and it really doesn't hurt in my opinion to have that implementation there in the core (given that it is lightweight and has no dependencies).

The Split-Reader API could be moved to flink-core, theoretically. I didn't argue for that in the original PR, because it is not used for simple sources (like here), but typically for sources with external systems that need separate I/O threads. These sources would be in separate modules anyways. So having the Split Reader API in a separate module would not make any difference to users, so I just went with your suggestion there (if it makes no difference, I typically go with what the code author suggests). IT has the slight advantage even. of being more flexible with test dependencies, so Integration Tests can be in the module directly.

### Packages in this PR

About the `lib` and `util` package names, I am not opinionated on those. In fact, I don't particularly like `lib` either.
Anything that is reasonably "discoverable" for users would work here from my side. Let's try to up with different names or locations, I am definitely open to suggestions.
Maybe we'll move this in a follow-up effort, to help unblock some features where I wanted to use this source in the tests.

**Review by becketqin — APPROVED** [review]
@StephanEwen Thanks a lot for the patient articulation. It is much clearer to me now. As you mentioned, we probably have to wait until 2.0 to get a complete cleanup of the APIs and package structure. While I prefer avoiding new tech debts, I don't think it is worth blocking other features, either. Therefore a followup refactor patch before 1.12 sounds a reasonable solution to me. +1 to merge the patch.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java
Comment (StephanEwen) [reply flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/NoSplitAvailableEvent.java:29]: That is right, we could remove them then.

File: flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java
Comment (StephanEwen) [reply flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/SplitRequestEvent.java:29]: I will do a follow-up where we consolidate these events.
I think they can be in `flink-core`, because they are just as much needed for sources using the Source API directly, rather than the SplitReader API.

File: flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java
Comment (StephanEwen) [reply flink-core/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceTest.java:39]: Fair enough. I'll add an ITCase for the NumberSequenceSource to cover the iterator sources.
Further iterator sources should then by fine with unit tests only.

**Comment by StephanEwen** [comment]
@becketqin  Here is the first followup, consolidating the events: https://github.com/apache/flink/pull/13533

**Pull request closed without merging**
